rasa/shared/core/events.py

```diff
@@ -507,17 +585,170 @@ def create_external(
         )
 
 
+# noinspection PyProtectedMember
+class DefinePrevUserUttered(Event):
+    """Defines the family of events that are used to update previous user utterance."""
+
+    def as_story_string(self) -> None:
+        """Returns the event as story string.
+
+        Returns:
+            None, as this event should not appear inside the story.
+        """
+        return
+
+
+# noinspection PyProtectedMember
+class DefinePrevUserUtteredFeaturization(DefinePrevUserUttered):
+    """Stores information whether action was predicted based on text or intent."""
```

Comment:
this sounds like something that should be stored on the action executed event and not on the user utterance 🤔

Reply:
but it is the property of user uttered. I think storing in the next action will be very confusing. Besides this information doesn't exist during training. `UserUttered` events are correct during training

Reply:
> action was predicted based on text or intent

this suggests to me that whether the intent or text is used depends on the used prediction algorithm which is used to predict the action. it seems counterintuitive to store such information on the source you are using for a prediction rather than the target, what if another algorithm wouldn't use the text to predict an action?

Reply:
> what if another algorithm wouldn't use the text to predict an action?
this is the thing, another algorithm should use this information. It is not part of the action, because during training it is part of user utterance